### PR TITLE
Derive a new max trip count for ONNXLoopOp when the upperbound, lowerbound and step are not constants

### PR DESCRIPTION
**Pull request by tungld**
Before this patch, we have a rewriting rule to derive a new max trip count for ONNXLoopOp when the upperbound, lowerbound and step are constants. In such case, the new max trip count is computed during the compile time.

This patch enhances the rule to support the case wherein the upperbound, lowerbound and step are not constants. In this case, we emit code to compute a derived trip count, and the new max trip count will be `min(old_max_trip_count, derived_max_trip_count)`.

This patch also adds several ops into OnnxBuilder and sorts the ops in the alphabet order.

### CODE REVIEW

**Comment by tungld** [comment]
@chentong319 this patch is a response to your second suggestion in the previous PR: https://github.com/onnx/onnx-mlir/pull/1449#pullrequestreview-981993732

File: src/Dialect/ONNX/DialectBuilder.cpp [hunk 1]
```diff
@@ -62,17 +72,26 @@ Value OnnxBuilder::matmul(Type Y, Value A, Value B, bool useGemm) const {
   return b.create<ONNXMatMulOp>(loc, Y, A, B);
 }
 
+Value OnnxBuilder::min(ValueRange inputs) const {
+  assert(inputs.size() >= 2 && "Expect at least two inputs");
```
Comment (etiotto) [inline src/Dialect/ONNX/DialectBuilder.cpp:76]: [nit]: could assert that all inputs have the same data type (using all_of or any_of).

File: src/Dialect/ONNX/DialectBuilder.cpp [hunk 2]
```diff
@@ -62,17 +72,26 @@ Value OnnxBuilder::matmul(Type Y, Value A, Value B, bool useGemm) const {
...
+  return b.create<ONNXMinOp>(loc, inputs[0].getType(), inputs);
+}
+
+Value OnnxBuilder::mul(Value A, Value B) const {
+  return b.create<ONNXMulOp>(loc, A, B);
```
Comment (etiotto) [inline src/Dialect/ONNX/DialectBuilder.cpp:81]: I think adding an assertion (here and for other binary operators) to check that the input have the same data type would be useful.

**Review by etiotto — APPROVED** [review]
LGTM - added a few suggestions.

File: src/Dialect/ONNX/Rewrite.cpp [hunk 3]
```diff
@@ -312,30 +311,87 @@ class LoopOpRewriteMaxTripCountPattern : public OpRewritePattern<ONNXLoopOp> {
...
+
+    // Case 1: the upper bound, lower bound and step are constants.
+    // - Compute the new max trip count at the compile time.
+    if (isDefinedByIntegerConstantOp(lbValue) &&
+        isDefinedByIntegerConstantOp(ubValue) &&
+        isDefinedByIntegerConstantOp(stepValue)) {
+      int64_t lowerBound = getOneIntergerConstant(lbValue);
+      int64_t upperBound = getOneIntergerConstant(ubValue);
+      int64_t step = getOneIntergerConstant(stepValue);
+      if ((step <= 0) || (upperBound - lowerBound) <= 0)
```
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/Rewrite.cpp:352]: nit `upperBound <= lowerBound` is easier to read. And if the trip count is zero? we could just get rid of the loop :-)

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM too

File: src/Dialect/ONNX/Rewrite.cpp [hunk 4]
```diff
@@ -312,30 +311,87 @@ class LoopOpRewriteMaxTripCountPattern : public OpRewritePattern<ONNXLoopOp> {
     Operation *addOp = cast<ONNXAddOp>(newCounterValue.getDefiningOp());
     Value counterValue = addOp->getOperands()[0];
     Value stepValue = addOp->getOperands()[1];
-    // 1. Step is a constant inside or outside the loop.
-    if (isInvariantArgConstant(stepValue, returnOp))
-      stepValue = getFedValue(stepValue, loopOp);
-    if (!isDefinedByIntegerConstantOp(stepValue))
-      return std::make_pair(false, -1);
-    // 2. Counter is an block argument and updated at each iteration.
+    // Counter is an block argument and updated at each iteration.
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Dialect/ONNX/Rewrite.cpp]: Done. Thanks!

File: src/Dialect/ONNX/DialectBuilder.cpp (hunk 2 above)
Comment (tungld) [reply src/Dialect/ONNX/DialectBuilder.cpp:98]: Done. Added an assertion for all binary operators.

File: src/Dialect/ONNX/DialectBuilder.cpp (hunk 1 above)
Comment (tungld) [reply src/Dialect/ONNX/DialectBuilder.cpp:82]: Done. Thanks!

**Pull request merged**